url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/render/jogl/JoglParametricCurveRenderer.java

```diff
@@ -0,0 +1,355 @@
+        gl.glVertex3d(vec.x + 0.02, vec.y - 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x + 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y - 0.02, vec.z);
+        gl.glEnd();
+
+        //p2
+        gl.glColor3d(colorTangPoint.r, colorTangPoint.g, colorTangPoint.b);
+        Vector3D vec2 = ptsB[tangPoint];
+        gl.glBegin(gl.GL_LINES);
+        gl.glVertex3d(vec.x + 0.02, vec.y - 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x + 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y - 0.02, vec.z);
+        gl.glEnd();
+
+        gl.glColor3d(colorLine.r, colorLine.g, colorLine.b);
+        gl.glBegin(gl.GL_LINE_STRIP);
+        gl.glVertex3d(vec.x, vec.y, vec.z);
+        gl.glVertex3d(vec2.x, vec2.y, vec2.z);
+        gl.glEnd();
+    }
+
+    static public void drawThreeControlPoints(GL2 gl, Vector3D[] ptsB,
+                                              ColorRgb colorLine,
+                                              ColorRgb colorCenterPoint,
+                                              ColorRgb colorTangPoint) {
+
+        gl.glColor3d(colorCenterPoint.r, colorCenterPoint.g, colorCenterPoint.b);
+        // p1
+        Vector3D vec = ptsB[0];
+        gl.glBegin(gl.GL_LINES);
+        gl.glVertex3d(vec.x + 0.02, vec.y - 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x + 0.02, vec.y + 0.02, vec.z);
+        gl.glVertex3d(vec.x - 0.02, vec.y - 0.02, vec.z);
+        gl.glEnd();
+
+        // p2
+        gl.glColor3d(colorTangPoint.r, colorTangPoint.g, colorTangPoint.b);
+        Vector3D vec2 = ptsB[1];
+        gl.glBegin(gl.GL_LINES);
+
+        gl.glVertex3d(vec2.x + 0.02, vec2.y - 0.02, vec2.z);
+        gl.glVertex3d(vec2.x - 0.02, vec2.y + 0.02, vec2.z);
+        gl.glVertex3d(vec2.x + 0.02, vec2.y + 0.02, vec2.z);
+        gl.glVertex3d(vec2.x - 0.02, vec2.y - 0.02, vec2.z);
+        gl.glEnd();
+
+        gl.glColor3d(colorLine.r, colorLine.g, colorLine.b);
+        gl.glBegin(gl.GL_LINE_STRIP);
+        gl.glVertex3d(vec.x, vec.y, vec.z);
+        gl.glVertex3d(vec2.x, vec2.y, vec2.z);
+        gl.glEnd();
+
+        gl.glColor3d(colorTangPoint.r, colorTangPoint.g, colorTangPoint.b);
+        vec2 = ptsB[2];
+        gl.glBegin(gl.GL_LINES);
+
+        gl.glVertex3d(vec2.x + 0.02, vec2.y - 0.02, vec2.z);
+        gl.glVertex3d(vec2.x - 0.02, vec2.y + 0.02, vec2.z);
+        gl.glVertex3d(vec2.x + 0.02, vec2.y + 0.02, vec2.z);
+        gl.glVertex3d(vec2.x - 0.02, vec2.y - 0.02, vec2.z);
+        gl.glEnd();
+
+        gl.glColor3d(colorLine.r, colorLine.g, colorLine.b);
+        gl.glBegin(gl.GL_LINE_STRIP);
+        gl.glVertex3d(vec.x, vec.y, vec.z);
+        gl.glVertex3d(vec2.x, vec2.y, vec2.z);
+        gl.glEnd();
+    }
+
+    public static void
+    drawTesselatedCurveInterior(GL2 gl, ParametricCurve curve)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 47 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.